File: python/beer-song/beer_song.py
text = """{0} bottle{2} of beer on the wall, {0} bottle{2} of beer.
Take {4} down and pass it around, {1} bottle{3} of beer on the wall.
"""

zero = """No more bottles of beer on the wall, no more bottles of beer.
Go to the store and buy some more, 99 bottles of beer on the wall.
"""

def verse(n):
    if n > 0:
        bottles1 = str(n)
        bottles2 = [str(n - 1), "no more"][n == 1]
        plural1 = ["s", ""][n == 1]
        plural2 = ["s", ""][n == 2]
        num = ["one", "it"][n == 1]
        return text.format(bottles1, bottles2, plural1, plural2, num)
    return zero


def song(start, end=0):
    out = ""
    for i in range(start, end-1, -1):
        out += verse(i)
        out += "\n"
    return out

File: fsharp/beer-song/BeerSong.fs
﻿module BeerSong

open System

let upper (str: string): string =
    match str |> List.ofSeq with
    | x :: xs -> (Char.ToUpper(x) :: xs) |> List.toArray |> String
    | _ -> str

let bottlesOfBeer (n: int): string = 
    match n with
    | 0 -> "no more bottles of beer"
    | 1 -> "1 bottle of beer"
    | _ -> sprintf "%i bottles of beer" n

let primary (n: int): string = 
    let str = bottlesOfBeer n
    sprintf "%s on the wall, %s." (upper str) str

let secondary (n: int): string =
    match n with
    | 0 -> "Go to the store and buy some more, 99 bottles of beer on the wall."
    | _ ->  
        let str = bottlesOfBeer (n - 1)
        let num = if n > 1 then "one" else "it"
        sprintf "Take %s down and pass it around, %s on the wall." num str

let couplet(start: int) (take: int) (n: int) : string list =
    let lines = [primary n; secondary n]
    if n + take - 1 = start then lines else lines @ [""]

let recite (start: int) (take: int): string list = 
    let range = [start .. -1 .. start - take + 1]
    range |> List.map (couplet start take) |> Seq.concat |> List.ofSeq
